gef.py

```diff
@@ -784,94 +789,118 @@ class Elf:
+            self.e_class, self.e_endianness = Elf.Class(e_class), Endianness(e_endianness)
 
-        # adjust endianness in bin reading
-        endian = gef.arch.endianness
+            if self.e_endianness != gef.arch.endianness:
+                warn("Unexpected endianness for architecture")
 
-        # off 0x7
-        self.e_osabi, self.e_abiversion = struct.unpack(f"{endian}BB", self.read(2))
+            endian = self.e_endianness
 
-        # off 0x9
-        self.e_pad = self.read(7)
+            # off 0x7
+            e_osabi, self.e_abiversion = self.read_and_unpack(f"{endian}BB")
+            self.e_osabi = Elf.OsAbi(e_osabi)
 
-        # off 0x10
-        self.e_type, self.e_machine, self.e_version = struct.unpack(f"{endian}HHI", self.read(8))
+            # off 0x9
+            self.e_pad = self.read(7)
 
-        # off 0x18
-        if self.e_class == Elf.ELF_64_BITS:
-            # if arch 64bits
-            self.e_entry, self.e_phoff, self.e_shoff = struct.unpack(f"{endian}QQQ", self.read(24))
-        else:
-            # else arch 32bits
-            self.e_entry, self.e_phoff, self.e_shoff = struct.unpack(f"{endian}III", self.read(12))
+            # off 0x10
+            e_type, e_machine, self.e_version = self.read_and_unpack(f"{endian}HHI")
+            self.e_type, self.e_machine = Elf.Type(e_type), Elf.Abi(e_machine)
 
-        self.e_flags, self.e_ehsize, self.e_phentsize, self.e_phnum = struct.unpack(f"{endian}IHHH", self.read(10))
-        self.e_shentsize, self.e_shnum, self.e_shstrndx = struct.unpack(f"{endian}HHH", self.read(6))
-
-        self.phdrs = []
-        for i in range(self.e_phnum):
-            self.phdrs.append(Phdr(self, self.e_phoff + self.e_phentsize * i))
+            # off 0x18
+            if self.e_class == Elf.Class.ELF_64_BITS:
+                self.e_entry, self.e_phoff, self.e_shoff = self.read_and_unpack(f"{endian}QQQ")
+            else:
+                self.e_entry, self.e_phoff, self.e_shoff = self.read_and_unpack(f"{endian}III")
 
-        self.shdrs = []
-        for i in range(self.e_shnum):
-            self.shdrs.append(Shdr(self, self.e_shoff + self.e_shentsize * i))
+            self.e_flags, self.e_ehsize, self.e_phentsize, self.e_phnum = self.read_and_unpack(f"{endian}IHHH")
+            self.e_shentsize, self.e_shnum, self.e_shstrndx = self.read_and_unpack(f"{endian}HHH")
 
-        if self.fd:
-            self.fd.close()
-            self.fd = None
+            self.phdrs : List["Phdr"] = []
+            for i in range(self.e_phnum):
+                self.phdrs.append(Phdr(self, self.e_phoff + self.e_phentsize * i))
 
+            self.shdrs : List["Shdr"] = []
+            for i in range(self.e_shnum):
+                self.shdrs.append(Shdr(self, self.e_shoff + self.e_shentsize * i))
         return
 
     def read(self, size):
         return self.fd.read(size)
 
+    def read_and_unpack(self, fmt):
```

Comment:
this is a nice helper! Keep it here for now but I could see this being pulled out into a function for all of gef to use.